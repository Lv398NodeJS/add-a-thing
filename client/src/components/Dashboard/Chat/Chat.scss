svg.icon {
  height: 1.5em;
}

.chat {
  &__container-backdrop {
    position: fixed;
    top: 0;
    left: 0;
    z-index: 10000;
    width: 100%;
    height: 100%;
    background-color: rgba(0, 0, 0, 0.6);
    transition: opacity 0.4s, transform 0.4s;
    cursor: pointer;

    &:not(.show) {
      opacity: 0;
      pointer-events: none;
    }
  }

  &__container {
    position: fixed;
    top: 0;
    right: 0;
    z-index: 10001;
    display: flex;
    flex-direction: column;
    width: 480px;
    max-width: 100%;
    height: 100%;
    background-color: #ffffff;
    transition: opacity 0.4s, transform 0.4s;

    &:not(.show) {
      transform: translateX(480px);
    }
  }

  &__header {
    flex-grow: 0;
  }

  &__messages {
    overflow-y: auto;
    flex-grow: 1;

    & a {
      color: #007bff;
      &:hover {
        text-decoration: underline;
      }
    }
  }

  &__input {
    flex-grow: 0;
  }
}
