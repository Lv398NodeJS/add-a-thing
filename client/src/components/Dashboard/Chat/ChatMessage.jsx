import React, { Component } from 'react';
import { Alert } from 'react-bootstrap';
import Linkify from 'linkifyjs/react';
import randomColor from 'random-material-color';

class ChatMessage extends Component {
  render() {
    const { text, date, userName } = this.props;

    const usernameColor = randomColor.getColor({
      text: userName,
      shades: ['800'],
    });
    const dateText = new Date(date)
      .toISOString()
      .replace('T', ' ')
      .replace(/:\d{2}\.\d{3}Z/, '');

    return (
      <Alert variant="secondary" className="p-1">
        <small className="float-right font-italic text-black-50">
          {dateText}
        </small>
        <b
          className="d-block"
          style={{
            color: usernameColor,
          }}
        >
          {userName}
        </b>
        <Linkify>
          {text}
        </Linkify>
      </Alert>
    );
  }
}

export default ChatMessage;
