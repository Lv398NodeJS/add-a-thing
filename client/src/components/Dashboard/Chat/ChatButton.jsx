import React, { Component } from 'react';
import {
  Button,
} from 'react-bootstrap';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import * as chatActions from '../../../actions/chatActions';
import './Chat.scss';
import { ReactComponent as MessageIcon } from './icons/message.svg';

class ChatButton extends Component {
  showChat = () => {
    const { chatActions: { setChatVisibility } } = this.props;
    setChatVisibility(true);
  };

  render() {
    return (
      <Button size="sm" variant="none" className="text-white p-0" onClick={this.showChat}>
        <MessageIcon className="icon" fill="#ffffff" />
      </Button>
    );
  }
}

const mapStateToProps = ({ chatReducer }) => ({
  ...chatReducer,
});
const mapDispatchToProps = dispatch => ({
  chatActions: bindActionCreators(chatActions, dispatch),
});

export { ChatButton as ChatButtonComponent };

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(ChatButton);
