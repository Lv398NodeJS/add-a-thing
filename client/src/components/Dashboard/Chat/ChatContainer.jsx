import React, { Component, createRef } from 'react';
import {
  Button, Alert,
  Container, FormControl, InputGroup,
} from 'react-bootstrap';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import * as chatActions from '../../../actions/chatActions';
import { sendMessage } from '../../../socket';
import './Chat.scss';
import { ReactComponent as SendIcon } from './icons/send.svg';

class ChatContainer extends Component {
  constructor(props) {
    super(props);
    const { chatActions: { setChatVisibility } } = this.props;
    this.textInputRef = createRef();
    this.messagesEndRef = createRef();

    setChatVisibility(false);
  }

  componentDidMount() {
    this.scrollToBottom();
  }

  componentDidUpdate() {
    this.scrollToBottom();
  }

  scrollToBottom = () => {
    const end = this.messagesEndRef.current;
    end.scrollIntoView({ behavior: 'smooth' });
  };

  hideChat = () => {
    const { chatActions: { setChatVisibility } } = this.props;
    setChatVisibility(false);
  };

  sendOnEnter = (event) => {
    if (event.key !== 'Enter') return;

    this.send();
  };

  send = () => {
    const input = this.textInputRef.current;
    const message = input.value.trim();
    sendMessage(message);
    input.value = '';
  };

  render() {
    const { visible, messages } = this.props;
    const messagesForRender = messages.map(message => (
      <Alert variant="secondary" className="p-1" key={message._id}>
        <small className="float-right font-italic text-black-50">
          {new Date(message.date).toLocaleString()}
        </small>
        {message.text}
      </Alert>
    ));

    return (
      <div>
        <div
          onClick={this.hideChat}
          className={['chat__container-backdrop', visible ? 'show' : 'hide'].join(' ')}
        />
        <Container
          fluid
          className={['chat__container', 'p-3', visible ? 'show' : 'hide'].join(' ')}
        >
          <div className="chat__header">
            <h4>Chat</h4>
          </div>
          <div className="chat__messages">
            {messagesForRender}
            <div ref={this.messagesEndRef} />
          </div>
          <div className="chat__input">
            <InputGroup>
              <FormControl
                onKeyPress={this.sendOnEnter}
                ref={this.textInputRef}
              />
              <InputGroup.Append>
                <Button variant="outline-primary" className="py-0" onClick={this.send}>
                  <SendIcon className="icon" />
                </Button>
              </InputGroup.Append>
            </InputGroup>
          </div>
        </Container>
      </div>
    );
  }
}

const mapStateToProps = ({ chatReducer }) => ({
  ...chatReducer,
});
const mapDispatchToProps = dispatch => ({
  chatActions: bindActionCreators(chatActions, dispatch),
});

export { ChatContainer as ChatContainerComponent };

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(ChatContainer);
