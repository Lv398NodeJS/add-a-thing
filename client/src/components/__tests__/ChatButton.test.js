import React from 'react';
import { render, shallow } from 'enzyme';
import { ChatButtonComponent } from '../Dashboard/Chat/ChatButton';

describe('<ChatButtonComponent />', () => {
  it('should render and match snapshot', () => {
    const testActions = {
      setChatVisibility: jest.fn(),
    };
    const renderTest = render(
      <ChatButtonComponent
        {...testActions}
      />,
    );
    expect(renderTest).toMatchSnapshot();
  });

  it('should show chat on click', () => {
    const testActions = {
      setChatVisibility: jest.fn(),
    };
    const renderTest = shallow(
      <ChatButtonComponent
        {...testActions}
      />,
    );
    renderTest.find('Button').simulate('click');
    expect(testActions.setChatVisibility).toHaveBeenCalledTimes(1);
    expect(testActions.setChatVisibility).toHaveBeenCalledWith(true);
  });
});
