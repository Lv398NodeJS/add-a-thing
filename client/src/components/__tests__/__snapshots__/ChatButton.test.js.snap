// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ChatButtonComponent /> should render and match snapshot 1`] = `
initialize {
  "0": Object {
    "attribs": Object {
      "class": "text-white btn btn-outline-primary btn-sm",
      "type": "button",
    },
    "children": Array [
      Object {
        "attribs": Object {
          "class": "icon",
          "fill": "#ffffff",
        },
        "children": Array [
          Object {
            "data": "message.svg",
            "next": null,
            "parent": [Circular],
            "prev": null,
            "type": "text",
          },
        ],
        "name": "svg",
        "namespace": "http://www.w3.org/2000/svg",
        "next": null,
        "parent": [Circular],
        "prev": null,
        "type": "tag",
        "x-attribsNamespace": Object {
          "class": undefined,
          "fill": undefined,
        },
        "x-attribsPrefix": Object {
          "class": undefined,
          "fill": undefined,
        },
      },
    ],
    "name": "button",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": null,
    "prev": null,
    "root": Object {
      "attribs": Object {},
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "namespace": "http://www.w3.org/1999/xhtml",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-attribsNamespace": Object {},
      "x-attribsPrefix": Object {},
    },
    "type": "tag",
    "x-attribsNamespace": Object {
      "class": undefined,
      "type": undefined,
    },
    "x-attribsPrefix": Object {
      "class": undefined,
      "type": undefined,
    },
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xml": false,
  },
}
`;
